libsplinter/src/circuit/mod.rs

```diff
@@ -380,94 +381,167 @@ impl<'r> IntoIterator for &'r Roster {
+                source: None,
+            })?;
+
+        Ok(circuit_directory.clone())
+    }
+
+    pub fn add_node(&mut self, id: String, node: SplinterNode) -> Result<(), SplinterStateError> {
+        {
+            let mut circuit_directory = self.circuit_directory.write().map_err(|_| {
+                SplinterStateError::new("Failed to get write guard for circuit directory".into())
+            })?;
+
+            circuit_directory.add_node(id, node);
+        }
+        self.commit_circuit_directory()?;
         Ok(())
     }
 
-    pub fn add_circuit(&mut self, name: String, circuit: Circuit) -> Result<(), WriteError> {
-        self.circuit_directory.add_circuit(name, circuit);
-        self.write_circuit_directory()?;
+    pub fn add_circuit(
+        &mut self,
+        name: String,
+        circuit: Circuit,
+    ) -> Result<(), SplinterStateError> {
+        {
+            let mut circuit_directory = self.circuit_directory.write().map_err(|_| {
+                SplinterStateError::new("Failed to get write guard for circuit directory".into())
+            })?;
+            circuit_directory.add_circuit(name, circuit);
+        }
+        self.commit_circuit_directory()?;
         Ok(())
     }
 
-    pub fn remove_node(&mut self, id: &str) -> Result<(), WriteError> {
-        self.circuit_directory.remove_node(id);
-        self.write_circuit_directory()?;
+    pub fn remove_node(&mut self, id: &str) -> Result<(), SplinterStateError> {
+        {
+            let mut circuit_directory = self.circuit_directory.write().map_err(|_| {
+                SplinterStateError::new("Failed to get write guard for circuit directory".into())
+            })?;
+            circuit_directory.remove_node(id);
+        }
+        self.commit_circuit_directory()?;
         Ok(())
     }
 
-    pub fn remove_circuit(&mut self, name: &str) -> Result<(), WriteError> {
-        self.circuit_directory.remove_circuit(name);
-        self.write_circuit_directory()?;
+    pub fn remove_circuit(&mut self, name: &str) -> Result<(), SplinterStateError> {
+        {
+            let mut circuit_directory = self.circuit_directory.write().map_err(|_| {
+                SplinterStateError::new("Failed to get write guard for circuit directory".into())
+            })?;
+            circuit_directory.remove_circuit(name);
+        }
+        self.commit_circuit_directory()?;
         Ok(())
     }
 
-    pub fn nodes(&self) -> &BTreeMap<String, SplinterNode> {
-        &self.circuit_directory.nodes()
+    pub fn nodes(&self) -> Result<BTreeMap<String, SplinterNode>, SplinterStateError> {
+        Ok(self.circuit_directory()?.nodes().clone())
     }
 
-    pub fn node(&self, node_id: &str) -> Option<&SplinterNode> {
-        self.circuit_directory.node(node_id)
+    pub fn node(&self, node_id: &str) -> Result<Option<SplinterNode>, SplinterStateError> {
+        Ok(self
+            .circuit_directory()?
```

Comment:
This results in a clone of the entire directory for every internal examination, which will be memory and time expensive as the number of circuits grow.  It should use the lock directly and return just a clone of the item being queried.  Same with the other usages